### PR TITLE
Add bindings nacos

### PR DESCRIPTION
**Pull request by cinience**
# Description

A binding that allows reading / writing from Nacos Configuration Service.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #810 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#1400

### CODE REVIEW

File: bindings/alicloud/nacos/nacos.go [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+	config := configParam{dataID: "", group: ""}
+	pair := strings.Split(s, ":")
+	config.dataID = strings.TrimSpace(pair[0])
+	if len(pair) == 2 {
+		config.group = strings.TrimSpace(pair[1])
+	}
+	if config.group == "" {
+		config.group = defaultGroup
+	}
+	if config.dataID == "" || config.group == "" {
```
Comment (pkedy) [inline bindings/alicloud/nacos/nacos.go:351]: Given the `if` above, would `config.group == ""` ever be true?

File: bindings/alicloud/nacos/nacos.go [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+			DataId: config.dataID,
+			Group:  config.group,
+		}); err != nil {
+			n.logger.Warnf("nacos cancel listener failed err: %v", err)
+		}
+	}
+}
+
+func (n *Nacos) notifyApp(group, dataID, content string) {
+	metadata := make(map[string]string, 2)
```
Comment (pkedy) [inline bindings/alicloud/nacos/nacos.go:310]: Simplify to
```golang
metadata := map[string]string{
	metadataConfigID:    dataID,
	metadataConfigGroup: group,
}
```

File: bindings/alicloud/nacos/nacos.go [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+		Group:    config.group,
+		Content:  "",
+		DatumId:  "",
+		OnChange: nil,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("fetch failed. err:%w", err)
+	}
+
+	if onchange := req.Metadata[metadataConfigOnchange]; onchange == "true" || onchange == "TRUE" {
```
Comment (pkedy) [inline bindings/alicloud/nacos/nacos.go:253]: Use `strings.EqualFold(onchange, "true")`?

File: bindings/alicloud/nacos/nacos.go [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+
+// Read implements InputBinding's Read method
+func (n *Nacos) Read(handler func(*bindings.ReadResponse) ([]byte, error)) error {
+	n.readHandler = handler
+
+	for _, watch := range n.watches {
+		go n.startListen(watch)
+	}
+
+	exitChan := make(chan os.Signal, 1)
```
Comment (pkedy) [inline bindings/alicloud/nacos/nacos.go:151]: I've seen this in a few components now and its less than ideal but the only way bindings can cleanup at the moment . Just pointing out this issue related to using a `Close() error` method for bindings. #779. I may implement this in `dapr/dapr` so you can use a context for cancelation.

File: bindings/alicloud/nacos/nacos.go [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+		go n.addListener4InputBinding(config)
+	}
+
+	return &bindings.InvokeResponse{Data: []byte(rst), Metadata: map[string]string{}}, nil
+}
+
+func (n *Nacos) addToWatches(c *configParam) bool {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	if n.watches != nil {
```
Comment (pkedy) [inline bindings/alicloud/nacos/nacos.go:263]: nitpick: You can safely range over a `nil` slice so this check is not needed.

File: bindings/alicloud/nacos/nacos_test.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+		atomic.AddInt32(&count, 1)
+
+		return nil, nil
+	}
+	go func() {
+		err = n.Read(handler)
+		require.NoError(t, err)
+	}()
+
+	time.Sleep(time.Second)
```
Comment (pkedy) [inline bindings/alicloud/nacos/nacos_test.go:61]: Use a channel here?

**Review by pkedy — CHANGES_REQUESTED** [review]
Two main changes I'd recommed:

1. Use `config.Decode`
2. Add a component level `ctx context.Context` and implement a `Close() error` func. See the Kafka pubsub as an example. I'll make calling `Close` from dapr/dapr work.

File: bindings/alicloud/nacos/nacos.go (hunk 4 above)
Comment (pkedy) [reply bindings/alicloud/nacos/nacos.go:143]: dapr/dapr#3088 make it so `Close() error` will be called on components. Please use a component level context for cancelation (see the Kafka pubsub for an example)

File: bindings/alicloud/nacos/nacos.go [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+	serverConfigs := make([]constant.ServerConfig, 0)
+	if ss == "" {
+		return serverConfigs, nil
+	}
+
+	array := strings.Split(ss, ",")
+	for _, s := range array {
+		cfg, err := parseServerURL(s)
+		if err != nil {
+			return serverConfigs, fmt.Errorf("%w", err)
```
Comment (pkedy) [inline bindings/alicloud/nacos/nacos.go:377]: Please include a message in `fmt.Errorf`

File: bindings/alicloud/nacos/nacos.go (hunk 4 above)
Comment (pkedy) [reply bindings/alicloud/nacos/nacos.go:143]: @cinience dapr/dapr#3088 was merged. Please implement `io.Closer` in this component.

**Review by pkedy — CHANGES_REQUESTED** [review]
See my comments re: dapr/dapr#3088.

**Comment by yaron2** [comment]
ping @pkedy

**Comment by pkedy** [comment]
@cinience I'm making a small tweaks that moves to of the decoding/validation into the metadata struct (which I renamed Settings). I'd like components to start moving toward this and this will be the first one!

One thing to note for the future is that the `convert` functions could have been implemented using `config.StringDecoder` and decoded directly into naco config types. I didn't want to make those kinds of big changes :)

Only really other thing to note is that I changed the mapstructure name for `TimeoutMs` to `timeout`.

**Review by pkedy — APPROVED** [review]
LGTM but of course did not actually test with nacos middleware. I recommend a follow up PR at somepoint to add an integration test for this.

**Review by pkedy — APPROVED** [review]

**Review by pkedy — APPROVED** [review]

**Review by artursouza — DISMISSED** [review]

**Review by artursouza — APPROVED** [review]

**Review by pkedy — APPROVED** [review]

**Comment by tanyo520** [comment]
When will the Nacos component be merged?

**Pull request merged**